README.md
FreenetIS CZ addresses database
=================

Requirements
------------
 - unzip
 - PHP 5.2.0
 - MySQL 5.1


Installation
------------
 - Create new database
 - Create database structure using db_structure.sql script
 - Run install.sh under root user
 - Run /var/www/freenetis-addresses/import.sh under root user for first time database update

Upgrade to 1.1.0
----------------
 - Before upgrading, execute SQL script in [upgrade_1.1.0.sql](upgrade_1.1.0.sql)
 
Changelog
---------
Changelog in debian format is available [here](changelog).


changelog
freenetis-address (1.1.0) stable; urgency=low
  * FOR UPGRADE INTORMATIONS READ README.md FIRST
  * Progress of preparing and importing is not printed in CRON mode
  * Server now return GPS position of address point
  * Server now searches in both town district and town qurters
 -- David Raska <jeffraska@gmail.com> Thu, 11 Sep 2014 16:00:00 +0200

freenetis-address (0.1.0) stable; urgency=low
  * First release
 -- David Raska <jeffraska@gmail.com> Thu, 03 Jul 2014 08:00:00 +0200

import.sh
#!/bin/bash

#
# This file is part of open source system FreenetIS
# and it is released under GPLv3 licence.
# 
# More info about licence can be found:
# http://www.gnu.org/licenses/gpl-3.0.html
# 
# More info about project can be found:
# http://www.freenetis.org/
# 
#


# init enviroment variable

function style
{
	if [[ $cron == FALSE ]]
	then
		tput $1 $2
	fi
}

function print_info
{
	echo -en "== "
	style bold
	echo -n $(date '+%D %T')
	style sgr0
	echo -n " == "
	style setaf 2
	style bold
	echo $1
	style sgr0
	echo
}

function print_warning
{
	echo -en "!! "
	style bold
	echo -n $(date '+%D %T')
	style sgr0
	echo -n " !! "
	style setaf 1
	style bold
	echo $1
	style sgr0
	echo
}

function control_c
{
	echo -e "\n"
	print_warning "SIGINT? OKAY :'("
	rm -rf $directory
	exit 1
}

echo "--------------------------------------------------------------------------------"

if [[ "$1" == "--cron" ]] || [[ "$1" == "-c" ]];
then
	cron=TRUE
else
	cron=FALSE
fi

if [ -r /etc/freenetis-addresses.ini ]
then
	. /etc/freenetis-addresses.ini
else
	print_warning "Config file not found"
	exit 1
fi

# do not modify
country_id=55	# ID of Czech republic
mysql_table="addresses"
mysql_table_old=$mysql_table"_old"
mysql_table_tmp=$mysql_table"_tmp"

# get start tim
starttime=$(date '+%s')

print_info "Trying to get latest database date"

# download web page
html=$(wget -O - http://nahlizenidokn.cuzk.cz/StahniAdresniMistaRUIAN.aspx 2>/dev/null)

if [ $? != 0 ];
then
	print_warning "Cannot get latest database date"
	rm -rf $directory
	exit 1
fi

# read date from html source
datestamp=$(echo $html | sed -nre "s/.*([0-9]{8})_OB_ADR_csv\.zip.*/\1/p")

if [ $? != 0 ];
then
	print_warning "Cannot get latest database date"
	rm -rf $directory
	exit 1
fi

db_datestamp=$(mysql $mysql_db -u $mysql_user -p$mysql_pass -h $mysql_server -P $mysql_port --silent -e "SELECT value FROM config WHERE name LIKE 'datestamp'")

if [ -n "$db_datestamp" ];
then
	if [ "$db_datestamp" -ge "$datestamp" ];
	then
		print_info "Database is up to date"
		exit 0
	else
		print_warning "Database is not up to date and will be updated"
	fi
else
	print_warning "Database is empty and will be imported"
fi

# create temp directory
directory=$(mktemp -d)

# catch SIGINT
trap control_c SIGINT

# download address database
print_info "Downloading address database"

wget http://vdp.cuzk.cz/vymenny_format/csv/$datestamp\_OB_ADR_csv.zip -O $directory/addresses.zip 2>/dev/null

if [ $? != 0 ];
then
	print_warning "Cannot download addresses"
	rm -rf $directory
	exit 1
fi

# unzip
print_info "Extracting address database"

unzip -qd $directory $directory/addresses.zip

if [ $? != 0 ];
then
	print_warning "Cannot extract addresses"
	rm -rf $directory
	exit 1
fi

# prepare database for importing
print_info "Preparing address database"

FILES=$directory/CSV/*.csv
FILE_NUM=$(ls -1 $FILES | wc -l)
I=0

for f in $FILES
do
	I=$(($I+1))
	if [[ $cron == FALSE ]]
	then
		echo -en "\rPreparing $I of $FILE_NUM: $f"
	fi

	#		remove columns				#change encoding					#create number with orientation number													#create number without orientaion number							#add country id
	cat $f | cut -s -d ";" -f 3-4,7-15 | iconv -f "WINDOWS-1250" -t "UTF-8" | sed -r 's/;(((č\.)(ev\.))|(č\.p\.));([0-9]*);([0-9][0-9]*);(.*)/;\4\3\6\/\7\8/g' | sed -r 's/;(((č\.)(ev\.))|(č\.p\.));([0-9]*);;/;\4\3\6/g' | sed -r "s/(.*)/$country_id;\1/g" > $f.utf8
	
	if [ $? != 0 ];
	then
		echo -e "\n"
		print_warning "Cannot prepare addresses - $I of $FILE_NUM: $f"
		rm -rf $directory
		exit 1
	fi
done

# import database
FILES=$directory/CSV/*.utf8
FILE_NUM=$(ls -1 $FILES | wc -l)
I=0

echo -e "\n"
print_info "Importing address database"

mysql $mysql_db -u $mysql_user -p$mysql_pass -h $mysql_server -P $mysql_port --silent -e "TRUNCATE TABLE $mysql_table_tmp"

if [ $? != 0 ];
then
	print_warning "Cannot clean temporary table"
	rm -rf $directory
	exit 1
fi

for f in $FILES
do
	I=$(($I+1))
	if [[ $cron == FALSE ]]
	then
		echo -en "\rImporting $I of $FILE_NUM: $f"
	fi

	mysql $mysql_db -u $mysql_user -p$mysql_pass -h $mysql_server -P $mysql_port --local-infile -e "LOAD DATA LOCAL INFILE '$f' INTO TABLE $mysql_table_tmp FIELDS TERMINATED BY ';' LINES TERMINATED BY '\n' IGNORE 1 LINES"
	
	if [ $? != 0 ];
	then
		echo -e "\n"
		print_warning "Cannot import addresses - $I of $FILE_NUM: $f"
		rm -rf $directory
		exit 1
	fi
done

echo -e "\n"
print_info "Updating database"

mysql $mysql_db -u $mysql_user -p$mysql_pass -h $mysql_server -P $mysql_port --silent -e "RENAME TABLE $mysql_table TO $mysql_table_old; RENAME TABLE $mysql_table_tmp TO $mysql_table; RENAME TABLE $mysql_table_old TO $mysql_table_tmp; REPLACE INTO config VALUES ('datestamp', '$datestamp'); TRUNCATE TABLE $mysql_table_tmp"

if [ $? != 0 ];
then
	print_warning "Cannot update database"
	rm -rf $directory
	exit 1
fi

# clean up
rm -rf $directory

# count script run time
endtime=$(date '+%s')
seconds=$(( $endtime - $starttime ))

hours=$(($seconds / 3600))
seconds=$(($seconds % 3600))
minutes=$(($seconds / 60))
seconds=$(($seconds % 60))

print_info "Import takes $(printf '%d:%02d:%02d\n' $hours $minutes $seconds)"


install.sh
#!/bin/bash

#
# This file is part of open source system FreenetIS
# and it is released under GPLv3 licence.
# 
# More info about licence can be found:
# http://www.gnu.org/licenses/gpl-3.0.html
# 
# More info about project can be found:
# http://www.freenetis.org/
# 
#

set -e

if [ "$(id -u)" != "0" ]; then
   echo "This script must be run as root" 1>&2
   exit 1
fi


if [[ "$1" == "-u" ]];
then
	echo "Uninstalling"
	rm -rf /var/www/freenetis-addresses
	rm -f /etc/freenetis-addresses.ini
	exit 1
fi

if [ ! -r /etc/freenetis-addresses.ini ]
then
	# get user name
	while [ true ]
	do
		read -p "DB user name: " user

		if [ -n "$user" ]
		then
			break
		fi
	done

	# get password
	read -sp "DB user password: " pass
	echo

	# get server
	read -p "DB server (default localhost): " server

	if [ -z "$server" ]
	then
		server="localhost"
	fi

	# get port
	read -p "DB port (default 3306): " port

	if [ -z "$port" ]
	then
		port="3306"
	fi

	# get database
	read -p "DB database (default addresses): " db

	if [ -z "$db" ]
	then
		db="addresses"
	fi

	echo "Saving configuration"

	# save configuration
	echo mysql_user=$user > /etc/freenetis-addresses.ini
	echo mysql_pass=$pass >> /etc/freenetis-addresses.ini
	echo mysql_server=$server >> /etc/freenetis-addresses.ini
	echo mysql_port=$port >> /etc/freenetis-addresses.ini
	echo mysql_db=$db >> /etc/freenetis-addresses.ini
fi

echo "Installing"

# install
rm -rf /var/www/freenetis-addresses
mkdir /var/www/freenetis-addresses

cp ./index.php /var/www/freenetis-addresses/
cp ./import.sh /var/www/freenetis-addresses/

chmod 700 /var/www/freenetis-addresses/import.sh

# configure CRON
echo "Preparing CRON"

echo "# /etc/cron.d/freenetis-addresses: Regular CRON file for freenetis-addressses (triggered each day)" > /etc/cron.d/freenetis-addresses
echo "" >> /etc/cron.d/freenetis-addresses
echo "SHELL=/bin/bash" >> /etc/cron.d/freenetis-addresses
echo $(shuf -i 0-59 -n 1) $(shuf -i 0-23 -n 1) "* * *   root    /var/www/freenetis-addresses/import.sh --cron >>\"/var/log/freenetis-addresses.log\" 2>&1" >> /etc/cron.d/freenetis-addresses

if [ -x /usr/sbin/invoke-rc.d ]; then
	invoke-rc.d cron restart 3>/dev/null || true
else
	/etc/init.d/cron restart 3>/dev/null || true
fi

exit 0


index.php
<?php
/*
 * This file is part of open source system FreenetIS
 * and it is released under GPLv3 licence.
 * 
 * More info about licence can be found:
 * http://www.gnu.org/licenses/gpl-3.0.html
 * 
 * More info about project can be found:
 * http://www.freenetis.org/
 * 
 */
 
function jstk2gps($x,$y,$h=200){
    /*Vypocet zemepisnych souradnic z rovinnych souradnic*/
	
	$a=6377397.15508; 
	$e=0.081696831215303;
	$n=0.97992470462083; 
	$konst_u_ro=12310230.12797036;
	$sinUQ=0.863499969506341; 
	$cosUQ=0.504348889819882;
	$sinVQ=0.420215144586493; 
	$cosVQ=0.907424504992097;
	$alfa=1.000597498371542; 
	$k=1.003419163966575;
	$ro=sqrt($x*$x+$y*$y);
	$epsilon=2*atan($y/($ro+$x));
	$D=$epsilon/$n; 
	$S=2*atan(exp(1/$n*log($konst_u_ro/$ro)))-M_PI_2;
	$sinS=sin($S);
	$cosS=cos($S);
	$sinU=$sinUQ*$sinS-$cosUQ*$cosS*cos($D);
	$cosU=sqrt(1-$sinU*$sinU);
	$sinDV=sin($D)*$cosS/$cosU; 
	$cosDV=sqrt(1-$sinDV*$sinDV);
	$sinV=$sinVQ*$cosDV-$cosVQ*$sinDV; 
	$cosV=$cosVQ*$cosDV+$sinVQ*$sinDV;
	$Ljtsk=2*atan($sinV/(1+$cosV))/$alfa;
	$t=exp(2/$alfa*log((1+$sinU)/$cosU/$k));
	$pom=($t-1)/($t+1);
	
	do
	{
		$sinB=$pom;
		$pom=$t*exp($e*log((1+$e*$sinB)/(1-$e*$sinB))); 
		$pom=($pom-1)/($pom+1);
	} 
	while (abs($pom-$sinB)>0.000000000000001);
	$Bjtsk=atan($pom/sqrt(1-$pom*$pom));
	
	/* Pravoúhlé souřadnice ve S-JTSK */   
	
	$a=6377397.15508; 
	$f_1=299.152812853;
	$e2=1-(1-1/$f_1)*(1-1/$f_1); 
	$ro=$a/sqrt(1-$e2*sin($Bjtsk)*sin($Bjtsk));
	$x=($ro+$H)*cos($Bjtsk)*cos($Ljtsk);  
	$y=($ro+$H)*cos($Bjtsk)*sin($Ljtsk);  
	$z=((1-$e2)*$ro+$H)*sin($Bjtsk);
		
	/* Pravoúhlé souřadnice v WGS-84*/
	
	$dx=570.69; 
	$dy=85.69; 
	$dz=462.84; 
	$wz=-5.2611/3600*M_PI/180;
	$wy=-1.58676/3600*M_PI/180;
	$wx=-4.99821/3600*M_PI/180; 
	$m=3.543*pow(10,-6); 
	$xn=$dx+(1+$m)*($x+$wz*$y-$wy*$z); 
	$yn=$dy+(1+$m)*(-$wz*$x+$y+$wx*$z); 
	$zn=$dz+(1+$m)*($wy*$x-$wx*$y+$z);
	
	/* Geodetické souřadnice v systému WGS-84*/
	
	$a=6378137.0; 
	$f_1=298.257223563;
	$a_b=$f_1/($f_1-1); 
	$p=sqrt($xn*$xn+$yn*$yn); 
	$e2=1-(1-1/$f_1)*(1-1/$f_1);
	$theta=atan($zn*$a_b/$p); 
	$st=sin($theta); 
	$ct=cos($theta);
	$t=($zn+$e2*$a_b*$a*$st*$st*$st)/($p-$e2*$a*$ct*$ct*$ct);
	$B=atan($t);  
	$L=2*atan($yn/($p+$xn));  
	$H=sqrt(1+$t*$t)*($p-$a/sqrt(1+(1-$e2)*$t*$t));
	
	/* Formát výstupních hodnot */   
	
	$B=$B/M_PI*180;				   
	//$sirka="N";
	if ($B<0)
	{
		$B=-$B; 
		//$sirka="S";
	}
	
	$stsirky=floor($B);  
	$B=($B-$stsirky)*60; 
	$minsirky=floor($B);
	$B=($B-$minsirky)*60; 
	$vtsirky=round($B*1000)/1000;
	//$sirka=$sirka+$stsirky+"°"+$minsirky+"'"+$vtsirky;
	$gps = $stsirky.'°'.$minsirky.'′'.$vtsirky.'″';
	
	$L=$L/M_PI*180;  
	//$delka="E";		 
	if ($L<0)
	{
		$L=-$L; 
		//$delka="W";
	}
	
	$stdelky=floor($L);  
	$L=($L-$stdelky)*60; 
	$mindelky=floor($L);
	$L=($L-$mindelky)*60; 
	$vtdelky=round($L*1000)/1000;
	
	$gps .= ' '.$stdelky.'°'.$mindelky.'′'.$vtdelky.'″';
		
	return $gps;
}

$config = parse_ini_file('/etc/freenetis-addresses.ini') or die();

define("mysql_user", $config['mysql_user']);
define("mysql_password", $config['mysql_pass']);
define("mysql_server", $config['mysql_server']);
define("mysql_port", $config['mysql_port']);
define("mysql_db", $config['mysql_db']);

// do not modify
define("mysql_table", 'addresses');

const STREET_NUMBER_MATCH_REGEX = '/((ev\.č\.)?[1-9][0-9]*(\/[1-9][0-9]*[a-zA-Z]*)*)$/m';

set_time_limit(1);

// get searched address
$country = @$_GET['country'];
$town = @$_GET['town'];
$district = @$_GET['district'];
$street = @$_GET['street'];
$zip = @$_GET['zip'];

$mode = @$_GET['mode'];

// connect do database
$db = mysql_connect(mysql_server.":".mysql_port, mysql_user, mysql_password) or die();
mysql_query("SET CHARACTER SET utf8", $db) or die();
mysql_query("SET NAMES utf8", $db) or die();
mysql_select_db(mysql_db, $db) or die();

$result = array();

if ($mode == 'test')
{
	$result = array(
			'state' => TRUE,
			'message' => 'Server is running' 
		);
}
else if ($mode == 'validate')
{
	if (!empty($country) && !empty($town) && !empty($zip) && !empty($street))
	{
		// find number
		$where = '';
		
		if (empty($district))
		{
			$district = $town;
		}
		
		$match = array();
		if (preg_match(STREET_NUMBER_MATCH_REGEX, $street, $match))
		{
			$street = trim(preg_replace(STREET_NUMBER_MATCH_REGEX, '', $street));
			
			// prepare query
			$query = "
				SELECT DISTINCT *
				FROM ".mysql_table."
				WHERE
					town_name LIKE '".mysql_real_escape_string($town)."' AND 
					zip_code LIKE '".mysql_real_escape_string($zip)."' AND
					street LIKE '".mysql_real_escape_string(trim($street))."' AND
					number LIKE '".mysql_real_escape_string($match[0])."' AND
					country = ".mysql_real_escape_string($country)." AND
					district_name LIKE '".mysql_real_escape_string($district)."'
				LIMIT 0,15";
			
			$mysql_result = mysql_query($query) or die(json_encode(array(mysql_error())));
			
			if (mysql_num_rows($mysql_result))
			{
				$result = array(
					'state' => TRUE,
					'message' => 'Address is valid'
				);
			}
			else
			{
				$result = array(
					'state' => FALSE,
					'message' => 'Address is not valid' 
				);
			}
		}
		else
		{
			$result = array(
				'state' => FALSE,
				'message' => 'Address is not valid' 
			);
		}
	}
	else
	{
		$result = array(
			'state' => FALSE,
			'message' => 'Address is not valid' 
		);
	}
}
else
{
	$where = '';
	
	if ($street !== NULL)
	{
		// find number
		$select = 'town_name, street, district_name';
		$orderby = '';

		$match = array();
		
		if (preg_match(STREET_NUMBER_MATCH_REGEX, $street, $match))
		{
			$street = trim(preg_replace(STREET_NUMBER_MATCH_REGEX, '', $street));
			if ($mode == 'gps')
			{
				$where = "AND number LIKE '".mysql_real_escape_string($match[0])."'";
			}
			else
			{
				$where = "AND number LIKE '".mysql_real_escape_string($match[0])."%'";
			}
			$select = '*';
			$orderby = "ORDER BY IF (number LIKE '".mysql_real_escape_string($match[0])."', 0, 1), number";
		}
		else if (strrpos($street, ' ') !== FALSE && 
				 strrpos($street, ' ') + 1 == strlen($street))
		{
			$select = '*';
		}

		if (!empty($town))
		{
			$where = "$where AND town_name LIKE '".mysql_real_escape_string($town)."'";
		}

		if (!empty($zip))
		{
			$where = "$where AND zip_code LIKE '".mysql_real_escape_string($zip)."'";
		}

		if (!empty($district))
		{
			$where = "$where AND district_name LIKE '".mysql_real_escape_string($district)."'";
		}
		
		if (!empty($country))
		{
			$where = "$where AND country = '".mysql_real_escape_string($country)."'";
		}

		// prepare query
		$query = "
			SELECT DISTINCT $select
			FROM ".mysql_table."
			WHERE street LIKE '".mysql_real_escape_string(trim($street))."%'
			$where
			$orderby
			LIMIT 0,15";
	}
	else if ($district !== NULL)
	{
		if (!empty($town))
		{
			$where = "$where AND town_name LIKE '".mysql_real_escape_string($town)."'";
		}
		
		if (!empty($country))
		{
			$where = "$where AND country = '".mysql_real_escape_string($country)."'";
		}

		$query = "
			SELECT DISTINCT district_name, town_name, town_quarter, IF ((district_name = town_name OR town_quarter LIKE '".mysql_real_escape_string($district)."') AND town_quarter NOT LIKE '', 1 , 0) AS same
			FROM ".mysql_table."
			WHERE (district_name LIKE '".mysql_real_escape_string($district)."%' OR town_quarter LIKE '".mysql_real_escape_string($district)."%')
			$where
			GROUP BY district_name, town_quarter
			ORDER BY same DESC, district_name ASC, town_quarter ASC
			LIMIT 0,15";
	}
	else if ($town !== NULL)
	{
		
		if (!empty($country))
		{
			$where = "AND country = '".mysql_real_escape_string($country)."'";
		}
		
		$query = "
			SELECT town_name, COUNT(district_name) as district_count
			FROM
			(
				SELECT town_name, district_name
				FROM ".mysql_table."
				WHERE town_name LIKE '".mysql_real_escape_string($town)."%'
				$where
				GROUP BY town_name, district_name
				ORDER BY town_name ASC
			) q
			GROUP BY town_name
			LIMIT 0,15";
	}

	// quit if no query
	if (!$query)
	{
		die(json_encode(array()));
	}
	
	// execute query
	$mysql_result = mysql_query($query) or die(json_encode(array(mysql_error())));

	// get results
	while ($row = mysql_fetch_assoc($mysql_result))
	{
		$result[] = $row;
	}
}

// send headers
@header('Cache-Control: no-cache, must-revalidate');
@header('Expires: Mon, 26 Jul 1997 05:00:00 GMT');

// send results
if ($mode == 'gps')
{
	@header('Content-type: text/plain; charset=utf-8');
	if (count($result) == 1)
	{
		// Count GPS coordinates from J-STK
		echo jstk2gps($result[0]['jstk_x'], $result[0]['jstk_y']);
	}
}
else
{
	@header('Content-type: application/json; charset=utf-8');
	echo json_encode($result);
}